python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py

```diff
@@ -0,0 +1,142 @@
+    MaterializeResult,
+    OpExecutionContext,
+)
+from dagster._annotations import public
+from dlt.common.pipeline import LoadInfo
+from dlt.extract.resource import DltResource
+
+if TYPE_CHECKING:
+    from dlt.extract.source import DltSource
+    from dlt.pipeline.pipeline import Pipeline
+
+from .constants import META_KEY_PIPELINE, META_KEY_SOURCE
+
+
+class DagsterDltResource(ConfigurableResource):
+    def _cast_load_info_metadata(self, mapping: Mapping[Any, Any]) -> Mapping[Any, Any]:
+        """Converts pendulum DateTime and Timezone values in a mapping to strings.
+
+        Workaround for dagster._core.errors.DagsterInvalidMetadata: Could not resolve the
+        metadata value for "jobs" to a known type. Value is not JSON serializable.
+
+        Args:
+            mapping (Mapping): dictionary possibly containing pendulum values
+
+        Returns:
+            mapping with pendulum DateTime and Timezone values casted to strings
+
+        """
+        from pendulum import DateTime, Timezone  # type: ignore
+
+        def _recursive_cast(value: Any):
+            if isinstance(value, dict):
+                return {k: _recursive_cast(v) for k, v in value.items()}
+            elif isinstance(value, list):
+                return [_recursive_cast(item) for item in value]
+            elif isinstance(value, (DateTime, Timezone)):
+                return str(value)
+            else:
+                return value
+
+        return {k: _recursive_cast(v) for k, v in mapping.items()}
+
+    def extract_resource_metadata(
+        self, resource: DltResource, load_info: LoadInfo
+    ) -> Mapping[str, Any]:
+        """Helper method to extract dlt resource metadata from load info dict.
+
+        Args:
+            resource (DltResource) dlt resource being materialized
+            load_info (LoadInfo) run metadata from dlt pipeline.run(...)
+
+        Returns:
+            Mapping[str, Any]: asset-specific metadata dictionary
+
+        """
+        dlt_base_metadata_types = {
+            "first_run",
+            "started_at",
+            "finished_at",
+            "dataset_name",
+            "destination_name",
+            "destination_type",
+        }
+
+        load_info_dict = self._cast_load_info_metadata(load_info.asdict())
+
+        # shared metadata that is displayed for all assets
+        base_metadata = {k: v for k, v in load_info_dict.items() if k in dlt_base_metadata_types}
+
+        # job metadata for specific target `resource.table_name`
+        base_metadata["jobs"] = [
+            job
+            for load_package in load_info_dict.get("load_packages", [])
+            for job in load_package.get("jobs", [])
+            if job.get("table_name") == resource.table_name
+        ]
+
+        return base_metadata
+
+    @public
+    def run(
+        self,
+        context: Union[OpExecutionContext, AssetExecutionContext],
+        **kwargs,
+    ) -> Iterator[Union[AssetMaterialization, MaterializeResult]]:
```

Comment:
I would recommend an intermediate object to be returned here so that we can potentially chain calls in the future.

This intermediate object would have similar functionality to `DbtCliInvocation`. call it `DltInvocation`. It could hold a reference to the `load_info` from the `dlt_pipeline` call if folks want to access that object for additional metadata to propagate into their Dagster events.

I'm fine for this to occur in a downstream PR, but I think this refactoring should be done before we expose this to the public.